### PR TITLE
[#809] Add test matrix for all different write configs

### PR DESCRIPTION
**Pull request by scottsand-db**
## Description
Add test matrix suite for all different ways to write to Delta for configs that do/do not start with `delta.` prefix. e.g. DataFrameWriter, DataStreamWriter, SQL API, etc.

## How was this patch tested?
Test suite.

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+                println(stmt)
+                // scalastyle:on println
+                sql(stmt)
+                sql("SHOW TBLPROPERTIES tbl").show(truncate = false)
+
+                val config = sql("SHOW TBLPROPERTIES tbl").collect().map { row =>
+                  val key = row.getString(0)
+                  val value = row.getString(1)
+                  (key, value)
+                }.toMap
```
Comment (zsxwing) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala:469]: Could you read from `Metadata.configuration` instead? `SHOW TBLPROPERTIES` may not output the same as `Metadata.configuration` IIRC.

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+  }
+
+  // scalastyle:off line.size.limit
+  /*
+  SQL Test Output
+  +---------------+-------------------------+--------------+---------+-------------------------+----------------------+--------------------------------+-----------------------------+
+  |Output Location|Config Input             |SQL Operation |AS SELECT|Contains OPTION no-prefix|Contains OPTION prefix|Contains TBLPROPERTIES no-prefix|Contains TBLPROPERTIES prefix|
+  +---------------+-------------------------+--------------+---------+-------------------------+----------------------+--------------------------------+-----------------------------+
+  |path           |options                  |create        |true     |false                    |true                  |false                           |false                        |
+  |path           |options                  |create        |false    |true                     |true                  |false                           |false                        |
```
Comment (zsxwing) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala:368]: This is the issue mentioned in #809 Right?

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala (hunk 1 above)
Comment (scottsand-db) [reply core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala]: done

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala (hunk 2 above)
Comment (scottsand-db) [reply core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala:376]: yes

**Review by zsxwing — COMMENTED** [review]
Thanks! Left a few minor comments.

Looks like we have two issues to fix:
- `createOrReplace` cannot overwrite a path with a different schema, but if we use the table name, it will work.
- We write `options.XXX` to Metadata.configuration.

**Comment by zsxwing** [comment]
@scottsand-db did you forget to push changes?

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+                tblBuilder.location(dir.getCanonicalPath).execute()
+                DeltaLog.forTable(spark, dir)
+              case ("table", _) =>
+                tblBuilder.tableName("tbl").execute()
+                DeltaLog.forTable(spark, TableIdentifier("tbl"))
+            }
+
+            log match {
+              case null =>
+                // CREATE OR REPLACE seems broken when using path and the table already exists
```
[the remark this answers is not in the document]
Comment (scottsand-db) [reply core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala:345]: yes

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+                }
+
+                // scalastyle:off println
+                println(stmt)
+                // scalastyle:on println
+
+                sql(stmt)
+
+                val log = DeltaLog.forTable(spark, TableIdentifier("tbl"))
+                val config = log.snapshot.metadata.configuration
```
[the remark this answers is not in the document]
Comment (scottsand-db) [reply core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala:492]: there are a bunch of different cases here ... as my latest PR update will show.

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+      }
+    }
+  }
+
+  /*
+  DeltaTableBuilder Test Output
+  +---------------+--------------+-------------------------------------+-------------------------+----------------------+-----+---------------------------------------------------------------------------------------+
+  |Output Location|Output Mode   |Contains No-Prefix Option (lowercase)|Contains No-Prefix Option|Contains Prefix-Option|ERROR|Config                                                                                 |
+  +---------------+--------------+-------------------------------------+-------------------------+----------------------+-----+---------------------------------------------------------------------------------------+
+  |path           |create        |true                                 |false                    |true                  |false|delta.deletedFileRetentionDuration -> interval 2 weeks,dataskippingnumindexedcols -> 33|
```
Comment (scottsand-db) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaWriteConfigsSuite.scala:297]: DeltaTableBuilder includes the non-prefix option but only lowercase @zsxwing

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaColumnRenameSuite.scala [hunk 6]
```diff
@@ -245,6 +245,7 @@ class DeltaColumnRenameSuite extends QueryTest
         spark.sql("alter table t1 rename column arr to arr1")
       }
 
+
```
Comment (zsxwing) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaColumnRenameSuite.scala:248]: nit: drop this change

**Review by zsxwing — APPROVED** [review]

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaColumnRenameSuite.scala (hunk 6 above)
Comment (scottsand-db) [reply core/src/test/scala/org/apache/spark/sql/delta/DeltaColumnRenameSuite.scala:248]: thanks.

Comment (scottsand-db) [reply core/src/test/scala/org/apache/spark/sql/delta/DeltaColumnRenameSuite.scala]: done

**Pull request closed without merging**